### ISSUE TITLE
Optimize HashBuild row storage by being join-type aware.

### ISSUE THREAD

**Issue by shiyu-bytedance** [facebookincubator/velox#6577]
### Description

In low cardinality Build-Side table cases, if the join is SEMI-variant or ANTI join, HashBuild side does not need to store all input rows, *indiscriminately* into the RowContainer before HashBuild step.

Instead, only rows with unique "join value"'s need to be stored, as the Join is used for "pruning" purpose.

Example:

Table A ANTI-JOIN Table B

| Table A      | 
| ----------- |
| 1      |
| 2   |
| 3  |

| Table B     | 
| ----------- |
| 1      |
| 1      |
| 2   |
| 2   |
| 3  |
| 3  |

Currently, HashBuild will store all 6 rows (1,1,2,2,3,3) before HashBuild. In big-data scenarios, this leads to OOM.

@mbasmanova @oerling

**Comment by mbasmanova** [comment]
@shiyu-bytedance This is a good point. HashBuild::setupTable has some logic to determine whether to allow duplicates or not. Is this not working properly?

**Comment by shiyu-bytedance** [comment]
@mbasmanova You are right, setupTable has dedup logic and works correctly. Although I believe that before rows are inserted into the hashtable, they are "bufferred" in RowContainer, which practically stores every row.

This behavior, that, before being hashed, *all* rows are stored temporarily, results in "peaky" ram usage. I wonder if an improvement is possible here; for example, "stream" input rows into the hashtable as they come in.

**Comment by mbasmanova** [comment]
> before being hashed, all rows are stored temporarily, results in "peaky" ram usage. 

This is a good point.

> I wonder if an improvement is possible here; for example, "stream" input rows into the hashtable as they come in.

Makes sense. Would you like to prototype something like this?

**Comment by shiyu-bytedance** [comment]
@mbasmanova 
Sure, happy to investigate and do a prototype for PR. Please assign this issue to me?

**Comment by mbasmanova** [comment]
Thank you, @shiyu-bytedance

**Comment by shiyu-bytedance** [comment]
@mbasmanova updated title to be more reflective of what this ticket is based on our discussion.

**Comment by liujiayi771** [comment]
We find the same issue in tpcds q14. There is a lot of duplicate key data in the left semi join, but the row container stores all the rows. @shiyu-bytedance Is there some progress so far?
